tachyon/zk/plonk/prover/synthesizer.h

```diff
@@ -0,0 +1,174 @@
+// Copyright 2022 The Halo2 developers
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
+// file.
+
+#ifndef TACHYON_ZK_PLONK_PROVER_SYNTHESIZER_H_
+#define TACHYON_ZK_PLONK_PROVER_SYNTHESIZER_H_
+
+#include <utility>
+#include <vector>
+
+#include "absl/numeric/bits.h"
+
+#include "tachyon/base/containers/container_util.h"
+#include "tachyon/zk/base/entities/prover.h"
+#include "tachyon/zk/plonk/constraint_system.h"
+#include "tachyon/zk/plonk/prover/witness_collection.h"
+
+namespace tachyon::zk {
+
+template <typename PCSTy>
+class Synthesizer {
+ public:
+  constexpr static size_t kMaxDegree = PCSTy::kMaxDegree;
+  constexpr static size_t kDomainSize = kMaxDegree + 1;
+
+  using F = typename PCSTy::Field;
+  using Poly = typename PCSTy::Poly;
+  using Evals = typename PCSTy::Evals;
+  using RationalEvals = typename PCSTy::RationalEvals;
+
+  Synthesizer() = default;
+  Synthesizer(size_t num_circuits, const ConstraintSystem<F>* constraint_system)
+      : num_circuits_(num_circuits), constraint_system_(constraint_system) {
+    advice_columns_vec_.resize(num_circuits_);
+    advice_blinds_vec_.resize(num_circuits_);
+    for (size_t i = 0; i < num_circuits_; ++i) {
+      // And these may be assigned with random order.
+      advice_columns_vec_[i] = base::CreateVector(
+          constraint_system->num_advice_columns(), Evals::Zero());
+      advice_blinds_vec_[i] = base::CreateVector(
+          constraint_system->num_advice_columns(), F::Zero());
+    }
+  }
+
+  // Synthesize circuit and store advice columns.
+  template <typename CircuitTy>
+  void GenerateAdviceValues(
+      Prover<PCSTy>* prover, const std::vector<CircuitTy>& circuits,
+      const std::vector<std::vector<Evals>>& instance_columns_vec) {
+    CHECK_EQ(num_circuits_, circuits.size());
+
+    for (Phase current_phase : constraint_system_->GetPhases()) {
+      for (size_t i = 0; i < num_circuits_; ++i) {
+        std::vector<RationalEvals> rational_advice_columns =
+            GenerateRationalAdvices(prover, current_phase,
+                                    instance_columns_vec[i], circuits[i]);
+
+        // Parse only indices related to the |current_phase|.
+        std::vector<size_t> indices_by_phase = base::FindIndices(
+            constraint_system_->advice_column_phases(), current_phase);
+
+        std::vector<Evals> evaluated_advice_columns =
+            SelectiveBatchEvaluate(indices_by_phase, rational_advice_columns);
+
+        for (size_t j = 0; j < evaluated_advice_columns.size(); ++j) {
+          // Add blinding factors to advice columns
+          *evaluated_advice_columns[j][kMaxDegree] = F(1);
+
+          CHECK(prover->CommitEvals(evaluated_advice_columns[j]));
+          SetAdvice(i, indices_by_phase[j],
+                    std::move(evaluated_advice_columns[j]),
+                    prover->blinder().Generate());
+        }
```

Comment:
I think you can save additional cost for allocating vector `evaluated_advice_columns`, here!

Reply:
```suggestion
        // Parse only indices related to the |current_phase|.
        const std::vector<Phase>& phases =
            constraint_system_->challenge_phases();
        for (size_t j = 0; j < phases.size(); ++i) {
          if (current_phase != phases[j]) continue;
          const RationalEvals& value = rational_advice_columns[j];
          std::vector<F> evaluated;
          CHECK(math::RationalField<F>::BatchEvaluate(value.evaluations(),
                                                      &evaluated));

          // Add blinding factors to advice columns
          *evaluated[kMaxDegree] = F::One();

          CHECK(prover->CommitEvals(evaluated));
          SetAdvice(i, j, std::move(evaluated), prover->blinder().Generate());
        }
```